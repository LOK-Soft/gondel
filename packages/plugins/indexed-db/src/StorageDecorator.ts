import { GondelComponent } from "@gondel/core";

/**
 * The @storage prop decorator will sync the selected value into a components index cache
 */
export function storage(version?: number) {
  return function<T extends GondelComponent>(target: T, propertyKey: string): void {
    const { _namespace, _componentName } = target;

    console.log(
      "storage binding to table %s_%s and item %s @v%s",
      _namespace,
      _componentName,
      propertyKey,
      version
    );
  };
}
