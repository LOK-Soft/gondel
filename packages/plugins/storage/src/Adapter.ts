export interface IStorage {
  readonly length: number;
  clear(): void;
  getItem(key: string): string | null;
  key(index: number): string | null;
  removeItem(key: string): void;
  setItem(key: string, value: string): void;
  [name: string]: any;
}

export interface ISerializer<T extends any = any> {
  serialize: (value: T) => string;
  deserialize: (value: string) => T;
}

type AdditionalAdapterConfiguration = {
  prefix: string;
  postfix: string;
  delimitier: string;
};

export class Adapter {
  private config: AdditionalAdapterConfiguration = {
    prefix: "",
    postfix: "",
    delimitier: ""
  };

  constructor(private name: string, private store: IStorage) {
    if (!store) {
      throw new Error(`Stroage not available in your environment`);
    }
  }

  configure(config: AdditionalAdapterConfiguration) {
    this.config = {
      ...this.config,
      ...config
    };
  }

  get<T, S extends ISerializer<T>>(key: string, serializer?: S): T | void {
    const value = this.store.getItem(this.generateAccessorKey(key));

    if (value) {
      if (serializer) {
        return serializer.deserialize(value);
      }

      return (value as unknown) as T;
    }

    return undefined;
  }

  set<T extends any = string>(key: string, value: T, serializer?: ISerializer<T>): void {
    let savableValue: any = "";

    if (serializer && typeof value !== "string") {
      savableValue = serializer.serialize(value);
    } else {
      savableValue = value;
    }

    this.store.setItem(this.generateAccessorKey(key), savableValue);
  }

  remove(key: string): void {
    this.store.removeItem(key);
  }

  clear(): void {
    this.store.clear();
  }

  toString(): string {
    return `Adapter { ${this.name} }`;
  }

  private generateAccessorKey(key: string): string {
    return [this.config.prefix, key, this.config.postfix].join(this.config.delimitier);
  }
}

export const localStorageAdapter = new Adapter("local", window.localStorage as IStorage);
export const sessionStorageAdapter = new Adapter("session", window.sessionStorage as IStorage);
