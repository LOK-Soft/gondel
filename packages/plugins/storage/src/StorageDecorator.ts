import { GondelComponent } from "@gondel/core";
import { Adapter, ISerializer, localStorageAdapter } from "./Adapter";

type GondelComponentWithStorage = GondelComponent & {
  // TODO: Optional props for third-party interaction (such as __gondelStorageBoundProps: [])?
};

type GondelComponentDecorator<T> = (target: T, propertyKey: string) => void;

type CreateStoreBindingsArgs<T> = {
  target: T;
  propertyKey: string;
  storeKey?: string;
  storageAdapter?: Adapter;
  serializer?: ISerializer;
};

// the default storage to use, better for switching whole projects
// to another storage type e.g. from local- to sessionstorage.
let internalDefaultStorageAdapter: Adapter = localStorageAdapter;

function createStorageBindings<T extends GondelComponentWithStorage>({
  target,
  propertyKey,
  storeKey = propertyKey,
  storageAdapter = internalDefaultStorageAdapter,
  serializer
}: CreateStoreBindingsArgs<T>): void {
  // TODO: should we use the component class name as prefix or sth?
  // const componentKey = target.constructor.name;
  Object.defineProperty(target, propertyKey, {
    get() {
      return storageAdapter.get(storeKey, serializer);
    },
    set(value: any) {
      storageAdapter.set(storeKey, value, serializer);
    }
  });
}

export const setDefaultStorageAdapter = (adapter: Adapter) =>
  (internalDefaultStorageAdapter = adapter);

export function storage<T extends GondelComponentWithStorage>(target: T, propertyKey: string): void;
export function storage<T extends GondelComponentWithStorage>(
  customAttributeKey: string,
  serializer?: ISerializer,
  storageAdapter?: Adapter
): GondelComponentDecorator<T>;
export function storage<T extends GondelComponentWithStorage>(
  targetOrStoreKey: T | string,
  propertyKeyOrSerializer: string | ISerializer | undefined,
  storageAdapter?: Adapter
): void | GondelComponentDecorator<T> {
  // First case will be used if we have a custom attribute and a valid serializer (which is typeof Serializer)
  if (typeof targetOrStoreKey === "string" && typeof propertyKeyOrSerializer !== "string") {
    const storeKey = targetOrStoreKey;
    const serializer = propertyKeyOrSerializer;

    return function<T extends GondelComponentWithStorage>(target: T, propertyKey: string): void {
      createStorageBindings({
        target,
        storageAdapter,
        storeKey,
        propertyKey,
        serializer
      });
    };
  }

  if (typeof targetOrStoreKey === "string" || typeof propertyKeyOrSerializer !== "string") {
    // this case should not occur, the only case could be a respec of the decorators
    throw new Error("Unexpected usage of @storage");
  }

  const target = targetOrStoreKey;
  const propertyKey = propertyKeyOrSerializer;

  // create simple bindings for @storage
  createStorageBindings({
    target,
    propertyKey
  });
}
