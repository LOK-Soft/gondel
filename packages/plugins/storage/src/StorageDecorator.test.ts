import {
  startComponents,
  getComponentByDomNode,
  Component,
  GondelBaseComponent,
  GondelComponent
} from "@gondel/core";
import { storage } from "./StorageDecorator";
import JSONSerializer from "./serializer/JSON";
import { Adapter } from "./Adapter";

function createMockElement(component: string, namespace: string = "g") {
  const buttonElement = document.createElement("div");
  buttonElement.setAttribute("data-" + namespace + "-name", component);
  document.documentElement!.appendChild(buttonElement);
  startComponents(buttonElement, namespace);
  return getComponentByDomNode(buttonElement, namespace)!;
}

beforeAll(() => {
  localStorage.clear();
  sessionStorage.clear();
});

const snapshotify = (val: any) => JSON.stringify(val);

describe("@gondel/plugin-data", () => {
  describe("@storage", () => {
    it("should bind to the localstorage", () => {
      @Component("Button")
      class Button extends GondelBaseComponent {
        @storage
        testProp: string;

        setTestProp() {
          this.testProp = "hellow there!";
        }
      }

      const button = createMockElement("Button") as Button;
      expect(button).toMatchSnapshot();
      expect(button.testProp).toBeUndefined();
      button.setTestProp();
      expect(button.testProp).toEqual("hellow there!");
      expect(button).toMatchSnapshot();
    });
  });

  describe("@storage()", () => {
    it("with serializer", () => {
      const tdata = {
        a: 10,
        b: "test",
        c: [1, "10", { d: [5] }]
      };

      @Component("Button")
      class Button extends GondelBaseComponent {
        @storage("anotherTestProp", JSONSerializer)
        testProp: Object;

        setTestProp() {
          this.testProp = tdata;
        }
      }

      const button = createMockElement("Button") as Button;
      expect(button).toMatchSnapshot();
      expect(button.testProp).toBeUndefined();
      button.setTestProp();
      expect(button.testProp).toEqual({
        a: 10,
        b: "test",
        c: [1, "10", { d: [5] }]
      });
      expect(button).toMatchSnapshot();
    });

    it("with custom adapter", () => {
      const customAdapter = new Adapter("custom-test", localStorage);
      customAdapter.configure({
        prefix: "custom",
        postfix: "custom",
        delimitier: "_"
      });

      customAdapter.set("test", "10");
      expect(snapshotify(localStorage)).toMatchSnapshot();
      expect(customAdapter.get("test")).toEqual("10");
      expect(localStorage.getItem("custom_test_custom")).toBeDefined();
    });

    it("syncs to all components on getter call", () => {
      @Component("A")
      class A extends GondelBaseComponent {
        @storage("sync")
        message: string;

        hello() {
          this.message = "Hello from A";
        }
      }

      @Component("B")
      class B extends GondelBaseComponent {
        @storage("sync")
        message: string;

        hello() {
          this.message = "Hello from B";
        }
      }

      const a = createMockElement("A") as A;
      const b = createMockElement("B") as B;
      expect(a).toMatchSnapshot();
      expect(b).toMatchSnapshot();
      expect(snapshotify(localStorage)).toMatchSnapshot();

      a.hello();
      expect(snapshotify(localStorage)).toMatchSnapshot();
      expect(a.message).toEqual("Hello from A");
      expect(b.message).toEqual("Hello from A");

      b.hello();
      expect(snapshotify(localStorage)).toMatchSnapshot();
      expect(a.message).toEqual("Hello from B");
      expect(b.message).toEqual("Hello from B");
    });
  });
});
