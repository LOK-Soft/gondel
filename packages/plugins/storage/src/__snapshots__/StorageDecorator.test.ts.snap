// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@gondel/plugin-data @storage should bind to the localstorage 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data @storage should bind to the localstorage 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data @storage() syncs to all components on getter call 1`] = `
A {
  "_componentName": "A",
  "_ctx": <div
    data-g-name="A"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data @storage() syncs to all components on getter call 2`] = `
B {
  "_componentName": "B",
  "_ctx": <div
    data-g-name="B"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data @storage() syncs to all components on getter call 3`] = `"{\\"testProp\\":\\"hellow there!\\",\\"anotherTestProp\\":\\"{\\\\\\"a\\\\\\":10,\\\\\\"b\\\\\\":\\\\\\"test\\\\\\",\\\\\\"c\\\\\\":[1,\\\\\\"10\\\\\\",{\\\\\\"d\\\\\\":[5]}]}\\",\\"custom_test_custom\\":\\"10\\"}"`;

exports[`@gondel/plugin-data @storage() syncs to all components on getter call 4`] = `"{\\"testProp\\":\\"hellow there!\\",\\"anotherTestProp\\":\\"{\\\\\\"a\\\\\\":10,\\\\\\"b\\\\\\":\\\\\\"test\\\\\\",\\\\\\"c\\\\\\":[1,\\\\\\"10\\\\\\",{\\\\\\"d\\\\\\":[5]}]}\\",\\"custom_test_custom\\":\\"10\\",\\"sync\\":\\"Hello from A\\"}"`;

exports[`@gondel/plugin-data @storage() syncs to all components on getter call 5`] = `"{\\"testProp\\":\\"hellow there!\\",\\"anotherTestProp\\":\\"{\\\\\\"a\\\\\\":10,\\\\\\"b\\\\\\":\\\\\\"test\\\\\\",\\\\\\"c\\\\\\":[1,\\\\\\"10\\\\\\",{\\\\\\"d\\\\\\":[5]}]}\\",\\"custom_test_custom\\":\\"10\\",\\"sync\\":\\"Hello from B\\"}"`;

exports[`@gondel/plugin-data @storage() with custom adapter 1`] = `"{\\"testProp\\":\\"hellow there!\\",\\"anotherTestProp\\":\\"{\\\\\\"a\\\\\\":10,\\\\\\"b\\\\\\":\\\\\\"test\\\\\\",\\\\\\"c\\\\\\":[1,\\\\\\"10\\\\\\",{\\\\\\"d\\\\\\":[5]}]}\\",\\"custom_test_custom\\":\\"10\\"}"`;

exports[`@gondel/plugin-data @storage() with serializer 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data @storage() with serializer 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data setDefaultStorageAdapter should set global defaults 1`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;

exports[`@gondel/plugin-data setDefaultStorageAdapter should set global defaults 2`] = `
Button {
  "_componentName": "Button",
  "_ctx": <div
    data-g-name="Button"
  />,
  "_namespace": "g",
  "stop": [Function],
}
`;
